resolwe/process/fields.py

```diff
@@ -0,0 +1,286 @@
+        """Validate field value."""
+        if self.required and value is None:
+            raise ValidationError("field is required")
+
+        if value is not None and self.choices is not None:
+            choices = [choice for choice, _ in self.choices]
+            if value not in choices:
+                raise ValidationError("field must be one of: {}".format(
+                    ", ".join(choices),
+                ))
+
+    def clean(self, value):
+        """Run validators and return the clean value."""
+        if value is None:
+            value = self.default
+
+        value = self.to_python(value)
+        try:
+            self.validate(value)
+        except ValidationError as error:
+            raise ValidationError("invalid value for {}: {}".format(
+                self.name,
+                error.args[0]
+            ))
+        return value
+
+    def __repr__(self):
+        return '<{klass} name={name} type={type} label="{label}">'.format(
+            klass=self.__class__.__name__,
+            name=self.name,
+            type=self.get_field_type(),
+            label=self.label,
+        )
+
+
+class StringField(Field):
+    field_type = 'basic:string'
+
+    def validate(self, value):
+        if value is not None and not isinstance(value, str):
+            raise ValidationError("field must be a string")
+
+        super().validate(value)
+
+
+class TextField(StringField):
+    field_type = 'basic:text'
+
+
+class BooleanField(Field):
+    field_type = 'basic:boolean'
+
+    def validate(self, value):
+        if value is not None and not isinstance(value, bool):
+            raise ValidationError("field must be a boolean")
+
+        super().validate(value)
+
+
+class IntegerField(Field):
+    field_type = 'basic:integer'
+
+    def to_python(self, value):
+        if value is not None:
+            return int(value)
+
+
+class FloatField(Field):
+    # TODO: Fix the underlying field into basic:float once that is renamed.
+    field_type = 'basic:decimal'
+
+    def to_python(self, value):
+        if value is not None:
+            return float(value)
+
+
+class DateField(Field):
+    field_type = 'basic:date'
+
+
+class DateTimeField(Field):
+    field_type = 'basic:datetime'
+
+
+class UrlField(StringField):
+    # Url types.
+    DOWNLOAD = 'download'
+    VIEW = 'view'
+    LINK = 'link'
+
+    URL_TYPES = (DOWNLOAD, VIEW, LINK)
+
+    def __init__(self, url_type, *args, **kwargs):
+        """Construct an URL field descriptor.
+
+        :param url_type: Type of URL
+        """
+        if url_type not in self.URL_TYPES:
+            raise ValueError('url_type must be one of: {}'.format(', '.join(self.URL_TYPES)))
+
+        self.url_type = url_type
+        super().__init__(*args, **kwargs)
+
+    def get_field_type(self):
+        return 'basic:url:{}'.format(self.url_type)
+
+
+class SecretField(Field):
+    field_type = 'basic:secret'
+
+
+class FileField(Field):
+    field_type = 'basic:file'
```

Comment:
This will require some thought. One thing that comes to mind is that `__repr__` and `__str__` could return an empty value.